### PR TITLE
Add Fragment Destructor which frees memory if it was malloc'd by the

### PR DESCRIPTION
**Pull request by eflumerf**
Fragment.

Reorder fields in FragmentHeader and TriggerRecordHeader to be 64-bit
aligned.

Update diagrams.

### CODE REVIEW

File: include/dataformats/TriggerRecordHeader.hpp [hunk 1]
```diff
@@ -27,16 +27,18 @@ struct TriggerRecordHeader
...
   timestamp_t trigger_record_end_time;
 
+  uint64_t n_requested_components; // NOLINT(build/unsigned)
+
+  run_number_t run_number;
   uint32_t error_bits; // NOLINT(build/unsigned)
 
-  uint64_t n_requested_components; // NOLINT(build/unsigned)
+  trigger_type_t trigger_type;
+  uint64_t unused : 48;
```
Comment (eflumerf) [inline include/dataformats/TriggerRecordHeader.hpp:41]: I'm actually not sure if this field has to be padded out to 64 bits or if another uint16_t would be sufficient to get it to a full 32-bit word...

**Review by philiprodrigues — APPROVED** [review]
All looks good to me

**Comment by bieryAtFnal** [comment]
Hi,
Hopefully, adding this comment to this Pull Request will get noticed, even though the PR has been closed...

When I updated my software development area to include the latest code on the dataformats main branch including this change, I noticed that the dataformats unit tests reported an error.

/afs/cern.ch/work/b/biery/public/dunedaq/current/sourcecode/dataformats/unittest/Fragment_test.cxx(50): info: check testFrag.get_trigger_timestamp() == 2 has passed
/afs/cern.ch/work/b/biery/public/dunedaq/current/sourcecode/dataformats/unittest/Fragment_test.cxx(51): info: check testFrag.get_run_number() == 3 has passed
*** Error in `/afs/cern.ch/work/b/biery/public/dunedaq/current/build/./dataformats/unittest/Fragment_test': double free or corruption (fasttop): 0x00000000009180d0 ***
======= Backtrace: =========
/lib64/libc.so.6(+0x81299)[0x7f78ff203299]
/afs/cern.ch/work/b/biery/public/dunedaq/current/build/./dataformats/unittest/Fragment_test[0x404b9b]
/afs/cern.ch/work/b/biery/public/dunedaq/current/build/./dataformats/unittest/Fragment_test[0x40466f]
/afs/cern.ch/work/b/biery/public/dunedaq/current/build/./dataformats/unittest/Fragment_test[0x403a6a]
/afs/cern.ch/work/b/biery/public/dunedaq/current/build/./dataformats/unittest/Fragment_test[0x4075aa]
/cvmfs/dune.opensciencegrid.org/dunedaq/DUNE/products/boost/v1_70_0/Linux64bit+3.10-2.17-e19-prof/lib/libboost_unit_test_framework.so.1.70.0(+0x33f99)[0x7f78ffe21f99]
/cvmfs/dune.opensciencegrid.org/dunedaq/DUNE/products/boost/v1_70_0/Linux64bit+3.10-2.17-e19-prof/lib/libboost_unit_test_framework.so.1.70.0(_ZN5boost17execution_monitor13catch_signalsERKNS_8functionIFivEEE+0x141)[0x7f78ffe20dd1]
/cvmfs/dune.opensciencegrid.org/dunedaq/DUNE/products/boost/v1_70_0/Linux64bit+3.10-2.17-e19-prof/lib/libboost_unit_test_framework.so.1.70.0(_ZN5boost17execution_monitor7executeERKNS_8functionIFivEEE+0x4f)[0x7f78ffe20e9f]
/cvmfs/dune.opensciencegrid.org/dunedaq/DUNE/products/boost/v1_70_0/Linux64bit+3.10-2.17-e19-prof/lib/libboost_unit_test_framework.so.1.70.0(_ZN5boost17execution_monitor8vexecuteERKNS_8functionIFvvEEE+0x54)[0x7f78ffe20f54]
/cvmfs/dune.opensciencegrid.org/dunedaq/DUNE/products/boost/v1_70_0/Linux64bit+3.10-2.17-e19-prof/lib/libboost_unit_test_framework.so.1.70.0(_ZN5boost9unit_test19unit_test_monitor_t21execute_and_translateERKNS_8functionIFvvEEEm+0x812)[0x7f78ffe4c752]
/cvmfs/dune.opensciencegrid.org/dunedaq/DUNE/products/boost/v1_70_0/Linux64bit+3.10-2.17-e19-prof/lib/libboost_unit_test_framework.so.1.70.0(+0x37eb2)[0x7f78ffe25eb2]
/cvmfs/dune.opensciencegrid.org/dunedaq/DUNE/products/boost/v1_70_0/Linux64bit+3.10-2.17-e19-prof/lib/libboost_unit_test_framework.so.1.70.0(+0x3816f)[0x7f78ffe2616f]
/cvmfs/dune.opensciencegrid.org/dunedaq/DUNE/products/boost/v1_70_0/Linux64bit+3.10-2.17-e19-prof/lib/libboost_unit_test_framework.so.1.70.0(+0x3816f)[0x7f78ffe2616f]
/cvmfs/dune.opensciencegrid.org/dunedaq/DUNE/products/boost/v1_70_0/Linux64bit+3.10-2.17-e19-prof/lib/libboost_unit_test_framework.so.1.70.0(_ZN5boost9unit_test9framework3runEmb+0x6b5)[0x7f78ffe287a5]
/cvmfs/dune.opensciencegrid.org/dunedaq/DUNE/products/boost/v1_70_0/Linux64bit+3.10-2.17-e19-prof/lib/libboost_unit_test_framework.so.1.70.0(_ZN5boost9unit_test14unit_test_mainEPFbvEiPPc+0x48b)[0x7f78ffe4a63b]
/afs/cern.ch/work/b/biery/public/dunedaq/current/build/./dataformats/unittest/Fragment_test[0x402d6f]
/lib64/libc.so.6(__libc_start_main+0xf5)[0x7f78ff1a4555]
/afs/cern.ch/work/b/biery/public/dunedaq/current/build/./dataformats/unittest/Fragment_test[0x402c29]
======= Memory map: ========
00400000-0040c000 r-xp 00000000 00:29 230838484                          /afs/cern.ch/work/b/biery/public/dunedaq/dfmodDev/build/dataformats/unittest/Fragment_test
0060b000-0060c000 r--p 0000b000 00:29 230838484                          /afs/cern.ch/work/b/biery/public/dunedaq/dfmodDev/build/dataformats/unittest/Fragment_test
0060c000-0060d000 rw-p 0000c000 00:29 230838484                          /afs/cern.ch/work/b/biery/public/dunedaq/dfmodDev/build/dataformats/unittest/Fragment_test
008fb000-0094e000 rw-p 00000000 00:00 0                                  [heap]
7f78f8000000-7f78f8021000 rw-p 00000000 00:00 0 
7f78f8021000-7f78fc000000 ---p 00000000 00:00 0 
7f78fed5e000-7f78fed75000 r-xp 00000000 fc:01 6382807                    /usr/lib64/libpthread-2.17.so
7f78fed75000-7f78fef74000 ---p 00017000 fc:01 6382807                    /usr/lib64/libpthread-2.17.so
7f78fef74000-7f78fef75000 r--p 00016000 fc:01 6382807                    /usr/lib64/libpthread-2.17.so
7f78fef75000-7f78fef76000 rw-p 00017000 fc:01 6382807                    /usr/lib64/libpthread-2.17.so
7f78fef76000-7f78fef7a000 rw-p 00000000 00:00 0 
7f78fef7a000-7f78fef81000 r-xp 00000000 fc:01 6382811                    /usr/lib64/librt-2.17.so
7f78fef81000-7f78ff180000 ---p 00007000 fc:01 6382811                    /usr/lib64/librt-2.17.so
7f78ff180000-7f78ff181000 r--p 00006000 fc:01 6382811                    /usr/lib64/librt-2.17.so
7f78ff181000-7f78ff182000 rw-p 00007000 fc:01 6382811                    /usr/lib64/librt-2.17.so
7f78ff182000-7f78ff346000 r-xp 00000000 fc:01 6382781                    /usr/lib64/libc-2.17.so
7f78ff346000-7f78ff545000 ---p 001c4000 fc:01 6382781                    /usr/lib64/libc-2.17.so
7f78ff545000-7f78ff549000 r--p 001c3000 fc:01 6382781                    /usr/lib64/libc-2.17.so
7f78ff549000-7f78ff54b000 rw-p 001c7000 fc:01 6382781                    /usr/lib64/libc-2.17.so
7f78ff54b000-7f78ff550000 rw-p 00000000 00:00 0 
7f78ff550000-7f78ff567000 r-xp 00000000 00:42 17559                      /cvmfs/dune.opensciencegrid.org/dunedaq/DUNE/products/gcc/v8_2_0/Linux64bit+3.10-2.17/lib64/libgcc_s.so.1
7f78ff567000-7f78ff766000 ---p 00017000 00:42 17559                      /cvmfs/dune.opensciencegrid.org/dunedaq/DUNE/products/gcc/v8_2_0/Linux64bit+3.10-2.17/lib64/libgcc_s.so.1
7f78ff766000-7f78ff767000 r--p 00016000 00:42 17559                      /cvmfs/dune.opensciencegrid.org/dunedaq/DUNE/products/gcc/v8_2_0/Linux64bit+3.10-2.17/lib64/libgcc_s.so.1
7f78ff767000-7f78ff768000 rw-p 00017000 00:42 17559                      /cvmfs/dune.opensciencegrid.org/dunedaq/DUNE/products/gcc/v8_2_0/Linux64bit+3.10-2.17/lib64/libgcc_s.so.1
7f78ff768000-7f78ff869000 r-xp 00000000 fc:01 6382789                    /usr/lib64/libm-2.17.so
7f78ff869000-7f78ffa68000 ---p 00101000 fc:01 6382789                    /usr/lib64/libm-2.17.so
7f78ffa68000-7f78ffa69000 r--p 00100000 fc:01 6382789                    /usr/lib64/libm-2.17.so
7f78ffa69000-7f78ffa6a000 rw-p 00101000 fc:01 6382789                    /usr/lib64/libm-2.17.so
7f78ffa6a000-7f78ffbde000 r-xp 00000000 00:42 17555                      /cvmfs/dune.opensciencegrid.org/dunedaq/DUNE/products/gcc/v8_2_0/Linux64bit+3.10-2.17/lib64/libstdc++.so.6.0.25unknown location(0): fatal error: in "Fragment_test/ExistingFragmentConstructor": signal: SIGABRT (application abort requested)
/afs/cern.ch/work/b/biery/public/dunedaq/current/sourcecode/dataformats/unittest/Fragment_test.cxx(51): last checkpoint
/afs/cern.ch/work/b/biery/public/dunedaq/current/sourcecode/dataformats/unittest/Fragment_test.cxx(29): Leaving test case "ExistingFragmentConstructor"; testing time: 863us
/afs/cern.ch/work/b/biery/public/dunedaq/current/sourcecode/dataformats/unittest/Fragment_test.cxx(19): Leaving test suite "Fragment_test"; testing time: 14383us
Leaving test module "Fragment_test"; testing time: 14401us

*** 1 failure is detected in the test module "Fragment_test"
End of unit test suite "/afs/cern.ch/work/b/biery/public/dunedaq/current/build/./dataformats/unittest/Fragment_test"

**Comment by eflumerf** [comment]
Fixed here: https://github.com/DUNE-DAQ/dataformats/pull/11

**Pull request merged**